### DISCUSSION TITLE
Forge cheatcode for explicit contract deployment address

### DISCUSSION THREAD

**Discussion by nicholascormier** [foundry-rs/foundry#4645] [Ideas]
A cheatcode to specify an address of a deployment inside of forge would be extremely useful imo. I have a contract (lets call this the "worker") that can only be called from another contract (the "controller"). Currently, ownership is hard-coded into the worker contract to save gas. This is where the bug was introduced, when a collaborator set this address to the deployment address of the controller that they observed on their machine, all of the tests broke on my machine. Since deployment addresses are not constant across environments, having a cheatcode to explicitly set the address of a deployment like `vm.deploy(<contract>, <address>)` or `vm.bind(<contract>, <address>)` would greatly improve developer experience as less configuration is needed between environments and less ambiguity when reasoning about "where" your contracts reside in forge.

**Comment by mds1** [comment]
There are solutions to your issue that don't require a new cheat code:
- Are you deploying with create2? If you, so you can set `bytecode_hash = 'none'` to ensure you get the same address on both machines (more info in the foundry book).
- Alternatively, instead of hardcoding the address as a constant, you can compute it and save it as an immutable (same gas costs). forge-std has helper methods for computing addresses with both create and create2

**Comment by nicholascormier** [comment]
Would this method not add unnecessary overhead when making production deployments? If the address making a mainnet deployment of the Controller is an EOA then the create2 calculation code inside of the Workers would just be for the sake of a working test environment and will have to be pruned to make the contract production ready. At that point, I think just an external method on the Worker to set the Controller address at runtime would be a less convoluted solution, but this again requires overhead that would be obsolete with a cheatcode like the one I described. I may just be completely misunderstanding something however as I am admittedly somewhat of a solidity newbie. Thanks for the help Matt.

**Reply by mds1** [reply]
Sorry, I'm not sure I fully understand the problem/question. There are various ways to get two contracts to know each other's addresses, such as be precomputing addresses or by having a setter method like you said. If you have a concrete example repo/script you can share maybe that would help

**Reply by nicholascormier** [reply]
The etch cheatcode is what I was looking for. Thanks for your help.
